sycl/include/sycl/ext/oneapi/experimental/non_uniform_groups.hpp

```diff
@@ -9,53 +9,63 @@
 #pragma once
 #include <CL/__spirv/spirv_ops.hpp>
 #include <CL/__spirv/spirv_vars.hpp>
+#include <sycl/types.hpp>
 #include <sycl/ext/oneapi/sub_group_mask.hpp>
 
 namespace sycl {
 __SYCL_INLINE_VER_NAMESPACE(_V1) {
-namespace ext::oneapi::experimental {
-
-template <class T> struct is_fixed_topology_group : std::false_type {};
-
-template <class T>
-inline constexpr bool is_fixed_topology_group_v =
-    is_fixed_topology_group<T>::value;
-
-#ifdef SYCL_EXT_ONEAPI_ROOT_GROUP
-template <> struct is_fixed_topology_group<root_group> : std::true_type {};
-#endif
-
-template <int Dimensions>
-struct is_fixed_topology_group<sycl::group<Dimensions>> : std::true_type {};
-
-template <> struct is_fixed_topology_group<sycl::sub_group> : std::true_type {};
 
-template <class T> struct is_user_constructed_group : std::false_type {};
-
-template <class T>
-inline constexpr bool is_user_constructed_group_v =
-    is_user_constructed_group<T>::value;
-
-#ifdef __SYCL_DEVICE_ONLY__
-// TODO: This may need to be generalized beyond uint32_t for big masks
 namespace detail {
-uint32_t CallerPositionInMask(sub_group_mask Mask) {
-  // FIXME: It would be nice to be able to jump straight to an __ocl_vec_t
+
+inline sycl::vec<unsigned, 4> ExtractMask(ext::oneapi::sub_group_mask Mask) {
   sycl::marray<unsigned, 4> TmpMArray;
   Mask.extract_bits(TmpMArray);
   sycl::vec<unsigned, 4> MemberMask;
   for (int i = 0; i < 4; ++i) {
     MemberMask[i] = TmpMArray[i];
   }
+  return MemberMask;
+}
+
+#ifdef __SYCL_DEVICE_ONLY__
+// TODO: This may need to be generalized beyond uint32_t for big masks
+inline uint32_t CallerPositionInMask(ext::oneapi::sub_group_mask Mask) {
+  sycl::vec<unsigned, 4> MemberMask = ExtractMask(Mask);
   auto OCLMask =
       sycl::detail::ConvertToOpenCLType_t<sycl::vec<unsigned, 4>>(MemberMask);
   return __spirv_GroupNonUniformBallotBitCount(
       __spv::Scope::Subgroup, (int)__spv::GroupOperation::ExclusiveScan,
       OCLMask);
 }
-} // namespace detail
 #endif
 
+template <typename NonUniformGroup>
+inline uint32_t IdToMaskPosition(NonUniformGroup Group, uint32_t Id) {
+  // TODO: This will need to be optimized
+  sycl::vec<unsigned, 4> MemberMask = ExtractMask(Group.Mask);
+  uint32_t Count = 0;
+  for (int i = 0; i < 4; ++i) {
+    for (int b = 0; b < 32; ++b) {
+      if (MemberMask[i] & (1 << b)) {
+        if (Count == Id) {
+          return i * 32 + b;
+        }
+        Count++;
+      }
+    }
+  }
+  __builtin_unreachable();
```

Comment:
Should this return an error? e.g. Id out of bounds/invalid etc?

Reply:
I don't think so.  This is device code, so we can't really throw a useful error.  The only way I can think of to signal an error would be to return some sort of sentinel value (e.g. always accessing lane 0 in the case of out-of-bounds behavior).

I thought that SYCL 2020 said this was UB, but it looks like that's not the case -- [the definition](https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html#_group_broadcast) of `group_broadcast` doesn't say anything about the range of acceptable local IDs.  I think this is a bug, though, because the restrictions are present for the "shuffles".  I'll open an issue against the SYCL spec.

Reply:
Opened https://github.com/KhronosGroup/SYCL-Docs/pull/390.